frontend/appflowy_tauri/src/appflowy_app/stores/effects/document/document_controller.ts

```diff
@@ -66,11 +67,116 @@ export class DocumentController {
     await this.backendService.applyActions(actions);
   };
 
+  getInsertAction = (node: Node, prevId: string | null) => {
+    return {
+      action: BlockActionTypePB.Insert,
+      payload: this.getActionPayloadByNode(node, prevId),
+    };
+  };
+
+  getUpdateAction = (node: Node) => {
+    return {
+      action: BlockActionTypePB.Update,
+      payload: this.getActionPayloadByNode(node, ''),
+    };
+  };
+
+  getMoveAction = (node: Node, parentId: string, prevId: string | null) => {
+    return {
+      action: BlockActionTypePB.Move,
+      payload: this.getActionPayloadByNode(
+        {
+          ...node,
+          parent: parentId,
+        },
+        prevId
+      ),
+    };
+  };
+
+  getDeleteAction = (node: Node) => {
+    return {
+      action: BlockActionTypePB.Delete,
+      payload: this.getActionPayloadByNode(node, ''),
+    };
+  };
+
   dispose = async () => {
     await this.backendService.close();
   };
 
+  private getActionPayloadByNode = (node: Node, prevId: string | null) => {
+    return {
+      block: this.getBlockByNode(node),
+      parent_id: node.parent || '',
+      prev_id: prevId || '',
+    };
+  };
+
+  private getBlockByNode = (node: Node) => {
+    return {
+      id: node.id,
+      parent_id: node.parent || '',
+      children_id: node.children,
+      data: JSON.stringify(node.data),
+      ty: node.type,
+    };
+  };
+
   private updated = (payload: Uint8Array) => {
-    console.log('didReceiveUpdate', payload);
+    const dispatch = this.dispatch;
+    if (!dispatch) return;
+    const { events, is_remote } = DocEventPB.deserializeBinary(payload);
+    if (!is_remote) return;
+    events.forEach((event) => {
+      event.event.forEach((_payload) => {
+        const { path, id, value, command } = _payload;
+        let valueJson;
+        try {
+          valueJson = JSON.parse(value);
+        } catch {
+          console.error('json parse error', value);
+          return;
+        }
+        if (!valueJson) return;
+
+        if (command === DeltaTypePB.Inserted || command === DeltaTypePB.Updated) {
+          // set map key and value ( block map or children map)
+          if (path[0] === 'blocks') {
+            if ('data' in valueJson && typeof valueJson.data === 'string') {
+              try {
+                valueJson.data = JSON.parse(valueJson.data);
+              } catch {
+                console.error('valueJson data parse error', valueJson.data);
+                return;
+              }
+            }
+            const block = {
```

Comment:
what about moving this `build` process to the block method. For example Block::from(value)